jerry-core/include/jerryscript-port.h

```diff
@@ -16,233 +16,301 @@
+ * `jerry_port_source_free`, which can be used to finalize the buffer.
  *
- * @param file_name_p Path that points to the EcmaScript file in the
- *                    filesystem.
- * @param out_size_p The opened file's size in bytes.
+ * @param file_name_p: Path that points to the source file in the filesystem.
+ * @param out_size_p: The opened file's size in bytes.
  *
- * @return the pointer to the buffer which contains the content of the file.
+ * @return pointer to the buffer which contains the content of the file.
  */
-uint8_t *jerry_port_read_source (const char *file_name_p, size_t *out_size_p);
+jerry_char_t *jerry_port_source_read (const char *file_name_p, jerry_size_t *out_size_p);
 
 /**
- * Frees the allocated buffer after the contents of the file are not needed
- * anymore.
+ * Free a source file buffer.
  *
- * @param buffer_p The pointer the allocated buffer.
+ * @param buffer_p: buffer returned by jerry_port_source_read
+ */
+void jerry_port_source_free (jerry_char_t *buffer_p);
+
+/**
+ * jerry-port-fs @}
+ */
+
+/**
+ * @defgroup jerry-port-date Date API
+ * @{
  */
-void jerry_port_release_source (uint8_t *buffer_p);
 
 /**
- * Default module resolver.
+ * Get local time zone adjustment in milliseconds for the given input time.
+ *
+ * The argument is a time value representing milliseconds since unix epoch.
+ *
+ * Ideally, this function should satisfy the stipulations applied to LocalTZA
+ * in section 21.4.1.7 of the ECMAScript version 12.0, as if called with isUTC true.
  *
- * Note:
- *      This port function is only used when JERRY_MODULE_SYSTEM is enabled.
+ * This port function can be called by jerry-core when JERRY_BUILTIN_DATE is enabled.
+ * Otherwise this function is not used.
  *
- * @param specifier Module specifier string.
- * @param referrer Parent module.
- * @param user_p An unused pointer.
+ * @param unix_ms: time value in milliseconds since unix epoch
  *
- * @return A module object if resolving is successful, an error otherwise.
+ * @return local time offset in milliseconds applied to UTC for the given time value
  */
-jerry_value_t jerry_port_module_resolve (const jerry_value_t specifier, const jerry_value_t referrer, void *user_p);
+int32_t jerry_port_local_tza (double unix_ms);
 
 /**
- * Release known modules.
+ * Get the current system time in UTC.
  *
- * Note:
- *      This port function should be called by the user application when
- *      the module database is no longer needed.
+ * This port function is called by jerry-core when JERRY_BUILTIN_DATE is enabled.
+ * It can also be used in the implementing application to initialize the random number generator.
  *
- * @param realm If this argument is object, release only those modules,
- *              which realm value is equal to this argument.
+ * @return milliseconds since Unix epoch
  */
-void jerry_port_module_release (const jerry_value_t realm);
+double jerry_port_current_time (void);
```

Comment:
Suggest return int64_t instead of `double`, there is no need  return `double`, as this is port

Reply:
According to https://tc39.es/proposal-temporal/docs/instant.html#epochNanoseconds

`epochNanoseconds` can be `bigint`, so maybe it's better to return nanoseconds instead,
as we using `ms` would lost precision.

Reply:
Temporal.Instant does represent time in nanosecond precision, however there's no constructor for Temporal.Instant that uses the current system time, it can only be constructed by providing a user specified nanosecond value, or from other objects.

From what I take from the proposal, it's that Temporal.Instant is used to handle external time values with higher precision, for example from a database, and not for querying system time.

I don't think a lot of platforms support true nanosecond resolution either. Unix uses microseconds, I beleive Windows supports even less. Embedded platfroms probably suport less as well.

Therefore we don't really need to have a more precise port api than the default Date object can handle. Also, we could not return with a bigint from the port, that would add a dependency to the core.

Whether we should use double or int64 is I think debatable. On one hand ecmascript requires Date objects to use double precision format. This is very unlikely to change. On the other, platforms usually represent time in integral types. This is also something that will likely never change.

We will need to convert from integer to double somewhere, question is whether it would be beneficial to do it later, and allow the port to return "true" system time.